### PR TITLE
Allowing GenomicsDBImport to skip header merging

### PR DESCRIPTION
**Pull request by lbergelson**
Adding an advanced option --sampleNameMap to GenomicsDBImport which can be specified instead of -V
* This allows providing a file mapping samples to their vcf uri which is used instead of loading each header and checking the sample names.
* The first file in the mapping is chosen to act as the header template.

If the headers are not consistent there may be data corruption.

A mapping file is a tab seperated file with no header in the format

Sample1	file1
Sample2 file2
Sample3 file3
...

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 1]
```diff
@@ -140,11 +125,23 @@
...
-                    "as the top Java layers has significantly higher overheads. This flag have no effect if only one " +
+                    "as the top Java layers have significantly higher overheads. This flag has no effect if only one " +
                     "batch is used. Defaults to false",
               optional = true)
     private Boolean doConsolidation = false;
 
+    @Advanced
+    @Argument(fullName = SAMPLE_NAME_MAP_LONG_NAME,
+            shortName = SAMPLE_NAME_MAP_LONG_NAME,
+            doc = "Path to file containing a mapping of sample name to file name in tab delimited format.  If this is " +
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: file name -> file URI?

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 2]
```diff
@@ -140,11 +125,23 @@
...
+                    "specified then the header from the first sample will be used as the header template rather than " +
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: used as the header template -> treated as the merged header rather than actually merging the headers

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 3]
```diff
@@ -87,34 +85,21 @@
     @Argument(fullName = StandardArgumentDefinitions.VARIANT_LONG_NAME,
               shortName = StandardArgumentDefinitions.VARIANT_SHORT_NAME,
               doc = "GVCF files to be imported to GenomicsDB. Each file must contain" +
-                    "data for only a single sample")
+                    "data for only a single sample. Either this or " + SAMPLE_NAME_MAP_LONG_NAME +
+                    " must be specified.",
+              optional = true,
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: We want this argument to be required though, if the sample name map is not provided

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 4]
```diff
@@ -140,11 +125,23 @@
...
+    @Argument(fullName = SAMPLE_NAME_MAP_LONG_NAME,
+            shortName = SAMPLE_NAME_MAP_LONG_NAME,
+            doc = "Path to file containing a mapping of sample name to file name in tab delimited format.  If this is " +
+                    "specified then the header from the first sample will be used as the header template rather than " +
+                    "merging the headers, and the sample names will taken from this file.  This is a performance optimization " +
+                    "that relaxes the normal checks for consistent headers.  Using vcfs with incompatible headers may result " +
+                    "in silent data corruption.",
+            optional = true,
+            mutex = {StandardArgumentDefinitions.VARIANT_LONG_NAME})
+    private String sampleNameToPathFile;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: Rename to `sampleNameMapFile`

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 5]
```diff
@@ -174,16 +171,16 @@ public int getDefaultCloudIndexPrefetchBufferSize() {
...
-    private Map<String, String> sampleNameToVcfUri = new LinkedHashMap<>();
+    private final Map<String, String> sampleNameToVcfUri = new LinkedHashMap<>();
 
     // Needed as smartMergeHeaders() returns a set of VCF header lines
-    private Set<VCFHeaderLine> mergedHeader = null;
+    private Set<VCFHeaderLine> mergedHeaderLines = null;
 
     // VCFHeader created from the merged header
-    private VCFHeader mergedVCFHeader;
+    private SAMSequenceDictionary mergedHeaderSequenceDictionary;
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: You forgot to update the comment for this field

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 6]
```diff
@@ -202,42 +199,98 @@ public int getDefaultCloudIndexPrefetchBufferSize() {
...
-        List<VCFHeader> headers = new ArrayList<>(variantPaths.size());
-
-        for (String variantPath : variantPaths) {
-
-            final AbstractFeatureReader<VariantContext, LineIterator> reader = getReaderFromVCFUri(variantPath);
-            VCFHeader header = (VCFHeader)reader.getHeader();
+    private void assertVariantPathsOrSampleNameFileWasSpecified(){
+        if ( (variantPaths == null || variantPaths.isEmpty()) && sampleNameToPathFile == null) {
+            throw new CommandLineException.MissingArgument(StandardArgumentDefinitions.VARIANT_LONG_NAME,
+                                                       "One of " + StandardArgumentDefinitions.VARIANT_LONG_NAME + " or " + SAMPLE_NAME_MAP_LONG_NAME + " must be specified" );
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: Insert "--" before arg names in error message for clarity

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 7]
```diff
@@ -202,42 +199,98 @@ public int getDefaultCloudIndexPrefetchBufferSize() {
...
+        super.onStartup();
+    }
 
-        List<VCFHeader> headers = new ArrayList<>(variantPaths.size());
-
-        for (String variantPath : variantPaths) {
-
-            final AbstractFeatureReader<VariantContext, LineIterator> reader = getReaderFromVCFUri(variantPath);
-            VCFHeader header = (VCFHeader)reader.getHeader();
+    private void assertVariantPathsOrSampleNameFileWasSpecified(){
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: It is unfortunate that our arg parsing system can't do this check for us.

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 8]
```diff
@@ -202,42 +199,98 @@ public int getDefaultCloudIndexPrefetchBufferSize() {
...
-            if (sampleNameToVcfUri.put(sampleName, variantPath) != null) {
-                throw new UserException("Duplicate sample " + sampleName);
+    private void initializeHeaderAndSampleMappings() {
+        // Only one of -V and -sampleNameToPathFile may be specified
+        if (sampleNameToPathFile == null) {
+            // -V was specified
+            final List<VCFHeader> headers = new ArrayList<>(variantPaths.size());
+
+            for (final String variantPath : variantPaths) {
+                final  VCFHeader header = loadHeaderFromVCFUri(variantPath);
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: Check that the header is non-null, throw if it is.

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 7 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: it is unfortunate.  I filed a bug https://github.com/broadinstitute/barclay/issues/52

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 9]
```diff
@@ -202,42 +199,98 @@ public int getDefaultCloudIndexPrefetchBufferSize() {
...
         }
+    }
 
-        mergedHeader = VCFUtils.smartMergeHeaders(headers, true);
-        mergedVCFHeader = new VCFHeader(mergedHeader);
+    private static void assertGVCFHasOnlyOneSample(String variantPath, VCFHeader header) {
+        // A GVCF file must contain only one sample, throw an exception otherwise
+        final int numberOfSamples = header.getNGenotypeSamples();
+        if (numberOfSamples != 1) {
+            throw new UserException("Input GVCF: " + variantPath + " was expected to contain a single sample but actually contained " + numberOfSamples + ".");
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: add "samples" at the end of this message

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 10]
```diff
@@ -202,42 +199,98 @@ public int getDefaultCloudIndexPrefetchBufferSize() {
...
+                }
             }
+            mergedHeaderLines = VCFUtils.smartMergeHeaders(headers, true);
+            mergedHeaderSequenceDictionary = new VCFHeader(mergedHeaderLines).getSequenceDictionary();
+        } else {
+            // -sampleNameMap was specified
+            sampleNameToVcfUri.putAll(loadSampleNameMapFile(IOUtils.getPath(sampleNameToPathFile)));
+            sampleNames.addAll(sampleNameToVcfUri.keySet());
+
+            final String firstHeaderPath = sampleNameToVcfUri.entrySet().iterator().next().getValue();
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: Is it guaranteed that the first line in the sample name map file corresponds to the first entry in `sampleNameToVcfUri`? Do you have an explicit unit test for this?

Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: I'd also add an explicit check here that the first entry matches the physical line in the map file, and throw a `GATKException` if it doesn't.

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 11]
```diff
@@ -174,16 +171,16 @@ public int getDefaultCloudIndexPrefetchBufferSize() {
     private List<ChromosomeInterval> intervals;
 
     // List of all sample names
-    private List<String> sampleNames = new ArrayList<>();
+    private final Set<String> sampleNames = new LinkedHashSet<>();
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: Maybe we should just use `keySet()` from `sampleNameToVcfUri` instead of maintaining this `Set` separately

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 12]
```diff
@@ -202,42 +199,98 @@ public int getDefaultCloudIndexPrefetchBufferSize() {
...
+    }
 
-        initializeIntervals();
-        super.onStartup();
+    @VisibleForTesting
+    static Map<String, String> loadSampleNameMapFile(final Path sampleToFileMapPath) {
+        try {
+            final Map<String, String> sampleToFilename = Files.lines(sampleToFileMapPath)
+                    .map(line -> {
+                        final String[] split = line.split("\\s");
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: Is it useful to require a single whitespace char of any type instead of a single tab? Also might have issues with newlines. Probably safest to just require a tab.

Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: Also I think you might want to pass in -1 as a second arg to split, to catch stray tabs at the end of a line

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 13]
```diff
@@ -202,42 +199,98 @@ public int getDefaultCloudIndexPrefetchBufferSize() {
...
+                        if (split.length != 2) {
+                            throw new UserException.BadInput(
+                                    "Expected a file of format\nSample\tFile\n but found line: " + line);
+                        }
+                        return split;
+                    })
+                    .collect(Collectors.toMap(split -> split[0],
+                                             split -> split[1],
+                                              (a, b) -> { throw new UserException.BadInput("Found two mappings for the same sample:\n" + a + "\n" + b); },
+                                              HashMap::new));
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: I assume this bug is fixed :)

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 14]
```diff
@@ -202,42 +199,98 @@ public int getDefaultCloudIndexPrefetchBufferSize() {
...
+
+            if (sampleToFilename.entrySet().isEmpty()) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: Can't you just write `sampleToFilename.isEmpty()` here?

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 15]
```diff
@@ -202,42 +199,98 @@ public int getDefaultCloudIndexPrefetchBufferSize() {
...
 
-        initializeIntervals();
-        super.onStartup();
+    @VisibleForTesting
+    static Map<String, String> loadSampleNameMapFile(final Path sampleToFileMapPath) {
+        try {
+            final Map<String, String> sampleToFilename = Files.lines(sampleToFileMapPath)
+                    .map(line -> {
+                        final String[] split = line.split("\\s");
+                        if (split.length != 2) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: Also check that both tokens are non-empty, non-whitespace

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 16]
```diff
@@ -247,29 +300,16 @@ public void onTraversalStart() {
 
         File workspaceDir = overwriteOrCreateWorkspace();
 
-        // If provided buffer size is less 1KB it is considered to be too small
-        // Either use default 16KB or any value larger than 10KB
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: Preserve this recommendation on the right value for `vcfBufferSizePerSample` in the docs for the argument above.

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 1 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 17]
```diff
@@ -355,8 +395,8 @@ public Object onTraversalSuccess() {
      * @param lowerSampleIndex  0-based Lower bound of sample index -- inclusive
      * @return  Feature readers to be imported in the current batch
      */
-    private Map<String, FeatureReader<VariantContext>> getFeatureReaders(List<String> sampleNames, Map<String, String> sampleNameToFileName, int batchSize, int sampleCount, int lowerSampleIndex) {
-        Map<String, FeatureReader<VariantContext>> sampleToReaderMap = new LinkedHashMap<>();
+    private Map<String, FeatureReader<VariantContext>> getFeatureReaders(final List<String> sampleNames, final Map<String, String> sampleNameToFileName, final int batchSize, int sampleCount, int lowerSampleIndex) {
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: Why are some of the int args not final?

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 18]
```diff
@@ -395,12 +435,12 @@ public Object onTraversalSuccess() {
...
      * @return  GenomicsDB import configuration object
      */
-    private GenomicsDBImportConfiguration.ImportConfiguration createImportConfiguration(
+    private static GenomicsDBImportConfiguration.ImportConfiguration createImportConfiguration(
         final String workspace,
         final String arrayName,
         final long variantContextBufferSize,
         final long segmentSize,
-        final long lbSampleIndex,
+        final  long lbSampleIndex,
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: Extra whitespace here

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 19]
```diff
@@ -202,42 +199,98 @@ public int getDefaultCloudIndexPrefetchBufferSize() {
...
+            mergedHeaderSequenceDictionary = new VCFHeader(mergedHeaderLines).getSequenceDictionary();
+        } else {
+            // -sampleNameMap was specified
+            sampleNameToVcfUri.putAll(loadSampleNameMapFile(IOUtils.getPath(sampleNameToPathFile)));
+            sampleNames.addAll(sampleNameToVcfUri.keySet());
+
+            final String firstHeaderPath = sampleNameToVcfUri.entrySet().iterator().next().getValue();
+            final VCFHeader header = loadHeaderFromVCFUri(firstHeaderPath);
+            mergedHeaderLines = header.getMetaDataInInputOrder();
+            mergedHeaderSequenceDictionary = header.getSequenceDictionary();
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: Should we throw if this header has no sequence dictionary? (which IS a possibility!)

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java [hunk 20]
```diff
@@ -202,42 +199,98 @@ public int getDefaultCloudIndexPrefetchBufferSize() {
...
+                final String sampleName = header.getGenotypeSamples().get(0);
+                sampleNames.add(sampleName);
+                final String previousPath = sampleNameToVcfUri.put(sampleName, variantPath);
+                if (previousPath != null) {
+                    throw new UserException("Duplicate sample: " + sampleName + ". Sample was found in both "
+                                                    + variantPath + " and " + previousPath + ".");
+                }
             }
+            mergedHeaderLines = VCFUtils.smartMergeHeaders(headers, true);
+            mergedHeaderSequenceDictionary = new VCFHeader(mergedHeaderLines).getSequenceDictionary();
```
Comment (droazen) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: Should we throw if the merged header has no sequence dictionary (theoretically possible)

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java [hunk 21]
```diff
@@ -29,6 +29,21 @@
...
+    private static final File TEST_REFERENCE_GENOME = new File(largeFileTestDir + "/Homo_sapiens_assembly38.20.21.fasta");
+
+    private static final String HG_00096 = largeFileTestDir + "gvcfs/HG00096.g.vcf.gz";
+    private static final String HG_00268 = largeFileTestDir + "gvcfs/HG00268.g.vcf.gz";
+    private static final String NA_19625 = largeFileTestDir + "gvcfs/NA19625.g.vcf.gz";
+    private static final List<String> LOCAL_GVCFS = Arrays.asList(HG_00096, HG_00268, NA_19625);
+
+    private static final String HG_00096_CLOUD = getGCPTestInputPath() + "large/gvcfs/HG00096.g.vcf.gz";
+    private static final String HG_00268_CLOUD = getGCPTestInputPath() + "large/gvcfs/HG00268.g.vcf.gz";
+    private static final String NA_19625_CLOUD = getGCPTestInputPath() + "large/gvcfs/NA19625.g.vcf.gz";
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java]: The tests were structured the way they were before precisely in order to avoid calling `getGCPTestInputPath()` outside of an `@Test` method (otherwise TestNG gives VERY hard-to-interpret error messages when it fails).

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java [hunk 22]
```diff
@@ -46,66 +61,73 @@ public String getTestedClassName() {
     }
 
     @Test
-    public void testGenomicsDBImportFileInputs() throws IOException {
-        final String hg00096 = largeFileTestDir + "gvcfs/HG00096.g.vcf.gz";
-        final String hg00268 = largeFileTestDir + "gvcfs/HG00268.g.vcf.gz";
-        final String na19625 = largeFileTestDir + "gvcfs/NA19625.g.vcf.gz";
-        final String combined = largeFileTestDir + "gvcfs/combined.gatk3.7.g.vcf.gz";
-        final SimpleInterval interval = new SimpleInterval("chr20", 17960187, 17981445);
+    public void testSampleMappingFileInsteadOfVCFs() throws IOException {
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java]: Move these tests after the "normal" tests for the tool

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java [hunk 23]
```diff
@@ -46,66 +61,73 @@ public String getTestedClassName() {
...
+        final String sampleFileContents = "HG00096\t" + HG_00096 +"\n" +
+                "HG00268\t"+ HG_00268 + "\n" +
+                "NA19625\t"+ NA_19625;
+        final File sampleNameFile = IOUtils.writeTempFile(sampleFileContents, "sampleNameMap", ".tx");
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java]: File extension here is weird

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java [hunk 24]
```diff
@@ -46,66 +61,73 @@ public String getTestedClassName() {
...
+        runCommandLine(args);
+        checkJSONFilesAreWritten(workspace);
+        checkGenomicsDBAgainstExpected(workspace, INTERVAL, COMBINED);
+    }
+
+    @Test(expectedExceptions = CommandLineException.class)
+    public void testCantSpecifyVCFAndSampleNameFile(){
+        final ArgumentsBuilder args = new ArgumentsBuilder()
+                .addArgument(GenomicsDBImport.SAMPLE_NAME_MAP_LONG_NAME, "someFile")
+                .addArgument(StandardArgumentDefinitions.VARIANT_LONG_NAME, "someotherfile")
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java]: Test would be more legit if these files actually existed (to prove that the error comes from specifying both args and not from file not found).

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java [hunk 25]
```diff
@@ -46,66 +61,73 @@ public String getTestedClassName() {
...
+    public void testGenomicsDBImportGCSInputsInBatches(final int batchSize) throws IOException {
+        testGenomicsDBImporterWithBatchSize(CLOUD_GVCFS, INTERVAL, COMBINED, batchSize);
     }
 
     /**
-     * Check whether user exception is thrown with vcf
-     * buffer size < 1024 bytes
      *
-     * @throws UserException  Value must be >1024 bytes
+     * @throws CommandLineException.OutOfRangeArgumentValue  Value must be >1024 bytes
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java]: > or >=?

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java [hunk 26]
```diff
@@ -131,7 +153,7 @@ private void testGenomicsDBImportWithZeroBufferSize(final List<String> vcfInputs
 
     }
 
-    private void writeToGenomicsDB(final List<String> vcfInputs, final SimpleInterval interval, final String workspace, int batchSize, final Boolean useBufferSize, int bufferSizePerSample) {
+    private void writeToGenomicsDB(final List<String> vcfInputs, final SimpleInterval interval, final String workspace, final int batchSize, final Boolean useBufferSize, int bufferSizePerSample) {
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java]: The last arg is not final?

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportUnitTest.java [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+
+    @DataProvider
+    public Object[][] getBadTestFiles(){
+        return new Object[][]{
+                {"Sample1\tsamplePath\n"
+                +"Sample1\tsamplePath"},  //duplicate sample name
+                {""}, // empty file
+                {"Sample1\tSample2\tFile"}, //various wrong numbers of inputs
+                {"Sample1\t"},
+                {"Sample1"}
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportUnitTest.java]: Also test:

* Empty first token, non-empty second token
* First token only whitespace
* Extra tab at end of otherwise normal line
* Sample and samplePath separated by a newline instead of a tab
* A line with only tabs

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportUnitTest.java [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+        GenomicsDBImport.loadSampleNameMapFile(sampleFile.toPath()  );
+    }
+
+    @Test
+    public void testSampleNameToFileMap(){
+        final File sampleFile = IOUtils.writeTempFile("Sample1\tfile1\nSample2\tfile2", "sampleMapping", ".txt");
+        final Map<String, String> expected = new HashMap<>();
+        expected.put("Sample1", "file1");
+        expected.put("Sample2", "file2");
+        Assert.assertEquals(GenomicsDBImport.loadSampleNameMapFile(sampleFile.toPath()), expected);
```
Comment (droazen) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportUnitTest.java]: Also assert that the first entry in the map returned is for `Sample1`

**Comment by droazen** [comment]
Review complete, back to @lbergelson

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 2 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 3 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: We have an integration test for that.  Limitation of the arg parser at the moment.

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 4 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 5 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: updated

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 6 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 8 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: guarded against

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 9 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 10 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: I've swithed everything to linked maps and sets.  That should fix the issue.

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 11 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: good idea, done

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 13 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: fixed

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 14 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: yes, I don't know why I didn't do that before...

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 15 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 12 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: done, and made it an ugly loop instead of a beautiful collections of millions of lambdas

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 17 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: because I had to scroll right to see them..

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 18 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: huh, i'm not sure what's going on here but I can't see the extra space in my editor, maybe intellij fixed it for me automagically

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 19 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: added a check

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 20 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: added a check

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java (hunk 21 above)
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java]: that's a very good point.  I will revert that and add a comment.

Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java]: I changed it to build these paths off of the local list at each test.  It's concise and can't fail statically.

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java (hunk 22 above)
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java]: moved

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java (hunk 23 above)
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java]: unweirded

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java (hunk 24 above)
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java]: @droazen  I don't think it's possible to get a file not found as a CommandLineParserException, but I see your point here.  I've updated to use the actual test files.

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java (hunk 25 above)
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java]: >=

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java (hunk 26 above)
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportIntegrationTest.java]: fixed

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportUnitTest.java (hunk 27 above)
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportUnitTest.java]: added, also added a test for whitespace within a token

File: src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportUnitTest.java (hunk 28 above)
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImportUnitTest.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java (hunk 18 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/genomicsdb/GenomicsDBImport.java]: nm, I was looking at the wrong whitespace

**Comment by lbergelson** [comment]
@droazen I think I've addressed all your comments.  Going to merge when tests pass.
